File: stdlib/lua/hull/tests/test_email.lua
-- test_email.lua - Tests for hull.email
--
-- Tests field validation and provider dispatch (no network I/O).
-- email.send follows the stdlib error convention: it THROWS a coded error
-- table ({ code, message }) on failure and returns true on success.
-- Run via: the C test harness (test_lua_runtime.c) loads and executes this.

local email = require('hull.email')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

-- Assert email.send(opts) throws a coded error whose code == want_code and
-- (optionally) whose message contains want_msg.
local function expect_code(name, opts, want_code, want_msg)
    test(name, function()
        local ok, err = pcall(email.send, opts)
        assert_eq(ok, false, "expected a throw")
        assert_eq(type(err), "table", "expected an error table")
        assert_eq(err.code, want_code, "code")
        if want_msg then
            assert(tostring(err):find(want_msg, 1, true),
                   "expected message to contain '" .. want_msg .. "', got: " .. tostring(err))
        end
    end)
end

-- ── validation ──────────────────────────────────────────────────────

expect_code("nil opts throws", nil, "invalid_argument", "opts required")

expect_code("missing from throws",
    { to = "x@y.com", subject = "s", body = "b" },
    "invalid_argument", "from required")

expect_code("missing to throws",
    { from = "x@y.com", subject = "s", body = "b" },
    "invalid_argument", "to required")

expect_code("missing subject throws",
    { from = "x@y.com", to = "y@z.com", body = "b" },
    "invalid_argument", "subject required")

expect_code("missing body throws",
    { from = "x@y.com", to = "y@z.com", subject = "s" },
    "invalid_argument", "body required")

expect_code("invalid from address throws",
    { from = "not-an-email", to = "y@z.com", subject = "s", body = "b" },
    "invalid_argument", "invalid from address")

expect_code("invalid to address throws",
    { from = "x@y.com", to = "not-an-email", subject = "s", body = "b" },
    "invalid_argument", "invalid to address")

-- ── provider dispatch ───────────────────────────────────────────────

expect_code("unknown provider throws",
    { provider = "unknown", from = "a@b.com", to = "c@d.com",
      subject = "s", body = "b" },
    "unknown_provider", "unknown provider")

test("default provider is smtp", function()
    -- Dispatches to the smtp adapter (which then fails because smtp is not
    -- configured / localhost is unreachable in the test env). The failure must
    -- NOT be unknown_provider - that would mean it never reached smtp.
    local ok, err = pcall(email.send, {
        from = "a@b.com", to = "c@d.com",
        subject = "s", body = "b",
        smtp_host = "localhost",
    })
    if not ok then
        assert(type(err) ~= "table" or err.code ~= "unknown_provider",
               "should dispatch to smtp, not fail with unknown_provider")
    end
end)

-- ── api provider validation ─────────────────────────────────────────

expect_code("postmark requires api_key",
    { provider = "postmark", from = "a@b.com", to = "c@d.com",
      subject = "s", body = "b" },
    "invalid_argument", "api_key required")

expect_code("sendgrid requires api_key",
    { provider = "sendgrid", from = "a@b.com", to = "c@d.com",
      subject = "s", body = "b" },
    "invalid_argument", "api_key required")

expect_code("resend requires api_key",
    { provider = "resend", from = "a@b.com", to = "c@d.com",
      subject = "s", body = "b" },
    "invalid_argument", "api_key required")

-- ── results ─────────────────────────────────────────────────────────

print(pass .. " passed, " .. fail .. " failed")
if fail > 0 then error(fail .. " test(s) failed") end

File: stdlib/js/hull/tests/test_email.js
// test_email.js — Tests for hull:email
//
// Tests field validation and provider dispatch (no network I/O).
// email.send follows the stdlib error convention: it THROWS an Error with a
// stable .code on failure and resolves to true on success.

import { email } from "hull:email";

let pass = 0;
let fail = 0;

async function test(name, fn) {
    try {
        await fn();
        pass++;
    } catch (e) {
        fail++;
        print("FAIL: " + name + ": " + (e && e.message));
    }
}

// Assert email.send(opts) throws a coded Error whose code === wantCode and
// (optionally) whose message contains wantMsg.
async function expectCode(name, opts, wantCode, wantMsg) {
    await test(name, async () => {
        let threw = null;
        try { await email.send(opts); }
        catch (e) { threw = e; }
        if (!threw) throw new Error("expected a throw");
        if (threw.code !== wantCode)
            throw new Error("expected code " + wantCode + ", got " + threw.code);
        if (wantMsg && String(threw).indexOf(wantMsg) === -1)
            throw new Error("expected message to contain '" + wantMsg + "', got: " + threw);
    });
}

await (async () => {
    // ── validation ──────────────────────────────────────────────────
    await expectCode("undefined opts throws", undefined, "invalid_argument", "opts required");
    await expectCode("null opts throws", null, "invalid_argument", "opts required");
    await expectCode("missing from throws",
        { to: "x@y.com", subject: "s", body: "b" }, "invalid_argument", "from required");
    await expectCode("missing to throws",
        { from: "x@y.com", subject: "s", body: "b" }, "invalid_argument", "to required");
    await expectCode("missing subject throws",
        { from: "x@y.com", to: "y@z.com", body: "b" }, "invalid_argument", "subject required");
    await expectCode("missing body throws",
        { from: "x@y.com", to: "y@z.com", subject: "s" }, "invalid_argument", "body required");
    await expectCode("invalid from address throws",
        { from: "bad", to: "y@z.com", subject: "s", body: "b" }, "invalid_argument", "invalid from address");
    await expectCode("invalid to address throws",
        { from: "x@y.com", to: "bad", subject: "s", body: "b" }, "invalid_argument", "invalid to address");

    // ── provider dispatch ───────────────────────────────────────────
    await expectCode("unknown provider throws",
        { provider: "unknown", from: "a@b.com", to: "c@d.com", subject: "s", body: "b" },
        "unknown_provider", "unknown provider");

    // ── api provider validation ─────────────────────────────────────
    await expectCode("postmark requires api_key",
        { provider: "postmark", from: "a@b.com", to: "c@d.com", subject: "s", body: "b" },
        "invalid_argument", "api_key required");
    await expectCode("sendgrid requires api_key",
        { provider: "sendgrid", from: "a@b.com", to: "c@d.com", subject: "s", body: "b" },
        "invalid_argument", "api_key required");
    await expectCode("resend requires api_key",
        { provider: "resend", from: "a@b.com", to: "c@d.com", subject: "s", body: "b" },
        "invalid_argument", "api_key required");

    // ── results ─────────────────────────────────────────────────────
    print(pass + " passed, " + fail + " failed");
    if (fail > 0) throw new Error(fail + " test(s) failed");
})();
